File: client/Scripts/Network/Protocol.cs
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace oily.top.Network
{
    // 与Java服务器完全匹配的消息协议
    public class ProtocolMessage
    {
        [JsonPropertyName("type")]
        public int Type { get; set; }

        [JsonPropertyName("data")]
        public JsonElement Data { get; set; }

        public ProtocolMessage() { }

        public ProtocolMessage(int type, object data)
        {
            Type = type;
            // 使用 System.Text.Json 序列化
            Data = JsonSerializer.SerializeToElement(data);
        }

        public string ToJson()
        {
            return JsonSerializer.Serialize(this);
        }

        public static ProtocolMessage FromJson(string json)
        {
            return JsonSerializer.Deserialize<ProtocolMessage>(json);
        }
    }

    // 消息类型（与Java服务器一致）
    public static class MessageType
    {
        public const int MSG_CONNECT_REQUEST = 1;
        public const int MSG_CONNECT_RESPONSE = 2;
        public const int MSG_PLAYER_MOVE = 3;
        public const int MSG_PLAYER_JOIN = 4;
        public const int MSG_PLAYER_LEAVE = 5;
        public const int MSG_WORLD_STATE = 6;
        public const int MSG_CHAT_MESSAGE = 7;
        public const int MSG_HEARTBEAT = 99;
    }

    // 连接请求
    public class ConnectRequest
    {
        [JsonPropertyName("playerName")]
        public string PlayerName { get; set; }

        public ConnectRequest() { }

        public ConnectRequest(string playerName)
        {
            PlayerName = playerName;
        }
    }

    // 连接响应
    public class ConnectResponse
    {
        [JsonPropertyName("playerId")]
        public int PlayerId { get; set; }

        [JsonPropertyName("x")]
        public float X { get; set; }

        [JsonPropertyName("y")]
        public float Y { get; set; }

        public ConnectResponse() { }

        public ConnectResponse(int playerId, float x, float y)
        {
            PlayerId = playerId;
            X = x;
            Y = y;
        }
    }

    // 玩家移动
    public class PlayerMove
    {
        [JsonPropertyName("playerId")]
        public int PlayerId { get; set; }

        [JsonPropertyName("x")]
        public float X { get; set; }

        [JsonPropertyName("y")]
        public float Y { get; set; }

        [JsonPropertyName("direction")]
        public float Direction { get; set; }

        public PlayerMove() { }

        public PlayerMove(int playerId, float x, float y, float direction)
        {
            PlayerId = playerId;
            X = x;
            Y = y;
            Direction = direction;
        }
    }

    // 玩家数据
    public class PlayerData
    {
        [JsonPropertyName("id")]
        public int Id { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("x")]
        public float X { get; set; }

        [JsonPropertyName("y")]
        public float Y { get; set; }

        [JsonPropertyName("direction")]
        public float Direction { get; set; }

        public PlayerData() { }

        public PlayerData(int id, string name, float x, float y, float direction)
        {
            Id = id;
            Name = name;
            X = x;
            Y = y;
            Direction = direction;
        }
    }

    // 世界状态
    public class WorldState
    {
        [JsonPropertyName("players")]
        public List<PlayerData> Players { get; set; }

        public WorldState() { }

        public WorldState(List<PlayerData> players)
        {
            Players = players;
        }
    }

    // 聊天消息
    public class ChatMessage
    {
        [JsonPropertyName("playerId")]
        public int PlayerId { get; set; }

        [JsonPropertyName("message")]
        public string Message { get; set; }

        public ChatMessage() { }

        public ChatMessage(int playerId, string message)
        {
            PlayerId = playerId;
            Message = message;
        }
    }
}

File: server/src/main/java/oily/top/network/Protocol.java
package oily.top.network;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.List;

public class Protocol {
    
    // 消息类型
    public static final int MSG_CONNECT_REQUEST = 1;
    public static final int MSG_CONNECT_RESPONSE = 2;
    public static final int MSG_PLAYER_MOVE = 3;
    public static final int MSG_PLAYER_JOIN = 4;
    public static final int MSG_PLAYER_LEAVE = 5;
    public static final int MSG_WORLD_STATE = 6;
    public static final int MSG_CHAT_MESSAGE = 7;
    public static final int MSG_HEARTBEAT=99;
    
    private static final ObjectMapper mapper = new ObjectMapper();
    
    static {
        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
    }
    
    public static class Message {
        public int type;
        public Object data;
        
        public Message() {}
        
        public Message(int type, Object data) {
            this.type = type;
            this.data = data;
        }
    }
    
    // 连接请求
    public static class ConnectRequest {
        public String playerName;
        
        public ConnectRequest() {}
        
        public ConnectRequest(String playerName) {
            this.playerName = playerName;
        }
    }
    
    // 连接响应
    public static class ConnectResponse {
        public int playerId;
        public float x;
        public float y;
        
        public ConnectResponse() {}
        
        public ConnectResponse(int playerId, float x, float y) {
            this.playerId = playerId;
            this.x = x;
            this.y = y;
        }
    }
    
    // 玩家移动
    public static class PlayerMove {
        public int playerId;
        public float x;
        public float y;
        public float direction;
        
        public PlayerMove() {}
        
        public PlayerMove(int playerId, float x, float y, float direction) {
            this.playerId = playerId;
            this.x = x;
            this.y = y;
            this.direction = direction;
        }
    }
    
    // 玩家数据
    public static class PlayerData {
        public int id;
        public String name;
        public float x;
        public float y;
        public float direction;
        
        public PlayerData() {}
        
        public PlayerData(int id, String name, float x, float y, float direction) {
            this.id = id;
            this.name = name;
            this.x = x;
            this.y = y;
            this.direction = direction;
        }
    }
    
    // 世界状态
    public static class WorldState {
        public List<PlayerData> players;
        
        public WorldState() {}
        
        public WorldState(List<PlayerData> players) {
            this.players = players;
        }
    }
    
    // 序列化
    public static String serialize(Message msg) throws JsonProcessingException {
        return mapper.writeValueAsString(msg);
    }
    
    // 反序列化
    public static Message deserialize(String json) throws JsonProcessingException {
        return mapper.readValue(json, Message.class);
    }
    
    // 创建消息
    public static Message createConnectResponse(int playerId, float x, float y) {
        return new Message(MSG_CONNECT_RESPONSE, new ConnectResponse(playerId, x, y));
    }
    
    public static Message createPlayerMove(int playerId, float x, float y, float direction) {
        return new Message(MSG_PLAYER_MOVE, new PlayerMove(playerId, x, y, direction));
    }
    
    public static Message createWorldState(List<PlayerData> players) {
        return new Message(MSG_WORLD_STATE, new WorldState(players));
    }
}